ray-curator/ray_curator/stages/audio/datasets/fleurs/create_initial_manifest.py

```diff
@@ -0,0 +1,134 @@
+
+    tar_gz_url = f"{base_lang_url}/audio/{split}.tar.gz"
+    urls.append(tar_gz_url)
+
+    return urls
+
+
+@dataclass
+class CreateInitialManifestFleursStage(ProcessingStage[_EmptyTask, SpeechObject]):
+    """
+    Stage to create initial manifest for the FLEURS dataset.
+
+    Dataset link: https://huggingface.co/datasets/google/fleurs
+
+    Will download all files, extract them, and create a manifest file with the
+    "audio_filepath" and "text" fields.
+
+    Args:
+        lang (str): Language to be processed, identified by a combination of ISO 639-1 and ISO 3166-1 alpha-2 codes.
+            Examples are:
+
+            - ``"hy_am"`` for Armenian
+            - ``"ko_kr"`` for Korean
+
+        split (str): Which dataset splits to process.
+            Options are:
+
+            - ``"test"``
+            - ``"train"``
+            - ``"dev"``
+
+        raw_data_dir (str): Path to the folder where the data archive should be downloaded and extracted.
+
+    Returns:
+        This srage generates an initial SpeechObject with the following fields:
+
+            {
+                "audio_filepath": <path to the audio file>,
+                "text": <transcription>,
+            }
+    """
+
+    lang: str
+    split: str
+    raw_data_dir: str
+    filepath_key: str = "audio_filepath"
+    text_key: str = "text"
+    _name: str = "CreateInitialManifestFleurs"
+
+    def process_transcript(self, file_path: str) -> list[SpeechObject]:
+        """
+        Parse transcript TSV file and put it inside manifest.
+        Assumes the TSV file has two columns: file name and text.
+        """
+
+        speech_tasks = []
+        root = os.path.dirname(file_path)
+        min_num_parts = 2  # Skip lines that don't have at least 2 parts
+
+        with open(file_path, encoding="utf-8") as fin:
+            for line in fin:
+                # Split the line into filename text using the tab delimiter
+                parts = line.strip().split("\t")
+                if len(parts) < min_num_parts:
+                    continue
+
+                file_name, transcript_text = parts[1], parts[2]
+                wav_file = os.path.join(root, file_name)
+
+                entry = {self.filepath_key: os.path.abspath(wav_file), self.text_key: transcript_text}
+                speech_task = SpeechObject(
+                    task_id=f"task_id_{file_path}",
+                    dataset_name=f"Fleurs_{self.lang}_{self.split}_{self.raw_data_dir}",
+                    filepath_key=self.filepath_key,
+                    data=entry,
+                )
+                speech_tasks.append(speech_task)
+
+        return speech_tasks
+
+    def download_extract_files(self, dst_folder: str) -> None:
+        """downloading and extracting files"""
+
+        os.makedirs(dst_folder, exist_ok=True)
+
+        # downloading all files
+        for file_url in get_fleurs_url_list(self.lang, self.split):
+            download_file(file_url, str(dst_folder))
```

Comment:
A bit surprised that you didn't run into some type of concurrency issue here. What if you made downloading its own stage?

cc @praateekmahajan

Reply:
I can also download and extract it in setup_on_node(), right?